applications/sarp-2-swdesign.md

```diff
@@ -0,0 +1,155 @@
+
+
+### Overview
+
+- **Total Estimated Duration:** 1 month
+- **Full-Time Equivalent (FTE):** 1 FTE
+- **Total Costs:** 30,000 USD
+
+
+### Milestone 1 Software Design & Bug Fixes
+
+- **Estimated duration:** 1 month
+- **FTE:**  1
+- **Costs:** 30,000 USD
+
+In our previous work, we found the following problems:
+
+1. **Crashes and timeouts of MIRAI** Certain pieces of substrate code lead to crashes of MIRAI. In other cases, parts of the code are not analyzed/do not produce warnings, because MIRAI runs into a timeout before reaching this code. Because of this, our examples are rather simple and we couldn't add and check tags at the locations we originally wanted to.
+
+2. **Complexity due to substrate macros** The main reason for crashes and timeouts in our examples, was caused by substrate macros, adding a lot of complexity to the code in the background. Ideally SARP only analyzes the newly written code of a pallet.
+
+3. **Invasiveness of tag analysis** The code we wrote in our PoC is very invasive and changes the code of the pallet. This is not practical for end-users. Ideally the user doesn't need to change anything on their code, or at least the changes should be very simple.
+
+To address 2. and 3. we plan to evaluate different software designs. These will be part of our deliverables and we plan to discuss these with Parity and/or Web3 Foundation.
+
+To address 1. we want to further analyze timeouts and crashes in MIRAI. Possibly they can be resolved by bugfixes in MIRAI. If not, we need to find workarounds.
+
+
+#### Deliverables
+
+| Number | Deliverable                  | Specification                                                                                                                                                  | 
+|--------|------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| 0a.    | License                      | MIT                                                                                                                                                            |
+| 0b.    | User Documentation           | We will provide a basic **tutorial** that explains how to reproduce our examples.                                                                              |
+| 0c.    | Testing and Testing Guide    | We will provide a testing guide on how to run SARP with the different software designs. These might be on different branches of the repository.                |
+| 1.     | Prototype Code               | Prototype code to showcase the different software designs.                                                                                                     | 
+| 2.     | Documentation                | Technical documentation describing the different variants of software design and its implications to pallet developers.                                        | 

```

Comment:
The most user-friendly approach is to generate MIRAI annotations automatically for certain bug classes like incorrect origins and inserting them in the right locations. Potentially, templates of annotations can also be provided for the more adventurous users to verify custom properties but this is not a priority.